### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by 1Dengaroo**

### CODE REVIEW

File: solution.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+xml_file = URI.open('http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml')
+xml_doc = Nokogiri::XML(xml_file)
+
+character_lines = Hash.new(0)
+
+xml_doc.xpath('//SPEECH').each do |speech|
+    speaker = speech.at_xpath('SPEAKER').text
+    if speaker == "ALL"
+        next
+    end
```
Comment (boniattirodrigo) [inline solution.rb:14]: Ruby has a shortcut for this.

```suggestion
    next if speaker == "ALL"
```

**Comment by boniattirodrigo** [comment]
Your solution is pretty a procedural code. Can you write a class for it and wrap some logic in methods?

**Comment by boniattirodrigo** [comment]
You can delete the macbeth.xml file.

File: solution.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    lines = speech.xpath('LINE')
+    character_lines[speaker] += lines.count
+end
+
+character_lines.sort_by {|name, num_lines| num_lines}.reverse.each do |name ,num_lines|
+    puts "#{name.downcase.capitalize} #{num_lines}"
+end
+
+
+xml_file.close
```
Comment (boniattirodrigo) [inline solution.rb:24]: Can you confirm this is needed even for a file open from the Internet?

Comment (1Dengaroo) [reply solution.rb]: I looked at the docs and I couldn't find any information specific to URLs from the internet. I found a website that suggested it was best practice to close it regardless of where I opened the file from, so I decided to keep it.

File: solution.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+class ShakespeareAnalyzer
+    attr_reader :character_lines
+
+    def initialize(url)
+        @url = url
+        @character_lines = Hash.new(0)
+    end
+
+    def analyze
+        character_lines.clear if not character_lines.empty?
```
Comment (boniattirodrigo) [inline solution.rb:13]: There are other ways to do this like:
```suggestion
        character_lines.clear if character_lines.any?
```
Or
```suggestion
        character_lines.clear unless character_lines.empty?
```

It's not too common to see `if not`

**Review by boniattirodrigo — APPROVED** [review]

**Pull request merged**